README
                             Muldis::D::RefEng
                                  README
---------------------------------------------------------------------------

by Darren Duncan <darren@DarrenDuncan.net>

This distribution features the Perl 5 module named Muldis::D::RefEng
(Reference Implementation of Muldis D Over Perl 5) and its support
structure.

TODO: Flesh this out.

Copyright © 2011-2015, Muldis Data Systems, Inc.


Changes
                             Muldis::D::RefEng
                                  Changes
---------------------------------------------------------------------------

2015-xx-xx

    * Muldis::D::RefEng version 0.0.0 for Perl 5 is released on CPAN as
    Muldis-D-RefEng-0.0.0.tar.gz.

    * As of this release, Muldis::D::RefEng is officially in pre-alpha
    development status.


lib/Muldis/D/RefEng.pm
use 5.008001;
use utf8;
use strict;
use warnings FATAL => 'all';

use Muldis::D::RefEng::Low_Level 0.000000;

###########################################################################
###########################################################################

{ package Muldis::D::RefEng; # package
    BEGIN {
        our $VERSION = '0.000000';
        $VERSION = eval $VERSION;
    }
} # package Muldis::D::RefEng

###########################################################################
###########################################################################

1; # Magic true value required at end of a reusable file's code.
__END__

=pod

=encoding utf8

=head1 NAME

Muldis::D::RefEng -
Reference Implementation of Muldis D Over Perl 5

=head1 VERSION

This document describes Muldis::D::RefEng version 0.0.0 for Perl 5.

=head1 SYNOPSIS

I<This documentation is pending.>

=head1 DESCRIPTION

Muldis::D::RefEng is the reference implementation of the Muldis D language,
authority B<http://muldis.com>, version number B<0.200.0>.  While other
Muldis D implementations also may exist, this one over Perl 5 is the
canonical one.

The separate all-documentation distribution L<Muldis::D> also exists; it
was intended to be the formal definition of the Muldis D language, but it
is currently years out of date; until it catches up, Muldis::D::RefEng also
serves double-duty as the formal language definition.

The Muldis D language itself has as a primary influence the work of Chris
Date (C.J. Date) and Hugh Darwen whose home website is
L<http://www.thethirdmanifesto.com/>.

I<This documentation is pending.>

=head1 INTERFACE

I<This documentation is pending.>

=head1 DIAGNOSTICS

I<This documentation is pending.>

=head1 CONFIGURATION AND ENVIRONMENT

I<This documentation is pending.>

=head1 DEPENDENCIES

This file requires any version of Perl 5.x.y that is at least 5.8.1.

=head1 INCOMPATIBILITIES

None reported.

=head1 SEE ALSO

These Perl 5 packages that are in the current distribution:
L<Muldis::D::RefEng::Low_Level>.

These other Perl 5 packages: L<Set::Relation>.

=head1 BUGS AND LIMITATIONS

I<This documentation is pending.>

=head1 AUTHOR

Darren Duncan - darren@DarrenDuncan.net

=head1 LICENSE AND COPYRIGHT

Copyright © 2011-2015, Muldis Data Systems, Inc.

L<http://www.muldis.com/>

Muldis::D::RefEng is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License (LGPL) as
published by the Free Software Foundation (L<http://www.fsf.org/>); either
version 3 of the License, or (at your option) any later version.  You
should have received a copy of the LGPL as part of the Muldis::D::RefEng
distribution, in the files named "LICENSE/LGPL" and "LICENSE/GPL" (the
LGPLv3 is defined as the terms of the GPLv3 plus extra permissions); if
not, see L<http://www.gnu.org/licenses/>.

If it is not feasible for you to employ Muldis::D::RefEng subject to the
terms of the LGPL, then the copyright holder of Muldis::D::RefEng can
provide you a customized proprietary license, often at no cost, so that it
is still possible for you to employ Muldis::D::RefEng to meet your needs.

Any versions of Muldis::D::RefEng that you modify and distribute must carry
prominent notices stating that you changed the files and the date of any
changes, in addition to preserving this original copyright notice and other
credits.  Muldis::D::RefEng is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  However, for an
additional fee, the copyright holders of Muldis::D::RefEng can sell you a
warranty for it.

While it is by no means required, the copyright holder of Muldis::D::RefEng
would appreciate being informed any time you create a modified version of
Muldis::D::RefEng that you are willing to distribute, because that is a
practical way of suggesting improvements to the standard version.

=head1 ACKNOWLEDGEMENTS

I<This documentation is pending.>

=head1 FORUMS

I<This documentation is pending.>

=cut


lib/Muldis/D/RefEng/StreamLexer.pm
use 5.008001;
use utf8;
use strict;
use warnings FATAL => 'all';

use Muldis::D::RefEng::StreamDecoder 0.000000;

###########################################################################
###########################################################################

my $MIN_TOKEN_FRAG_LEN =  10;  # in characters
my $DEF_TOKEN_FRAG_LEN = 100;  # in characters

###########################################################################
###########################################################################

{ package Muldis::D::RefEng::StreamLexer; # class
    BEGIN {
        our $VERSION = '0.000000';
        $VERSION = eval $VERSION;
    }

    # ATTRIBUTE LIST:
        # max_frag_len : Int : Max length for an output token or fragment.
        # char_stream : A filehandle open for reading, we pull chars from.
        # char_stream_buf : Str : Any chars we pulled from char_stream that
            # weren't yet output as a token or fragment; these are
            # conceptually un-pulled chars but we're using them to
            # effectively look ahead into the source stream.
            # We sometimes use char_stream_buf as active working memory.
        # is_mid_singquot_str : Bool : 
        # is_mid_doubquot_str : Bool : 
        # is_mid_backtick_str : Bool : 
        # is_mid_slashtar_str : Bool : 
        # is_mid_token : Bool : True iff we have yet to output the last
            # fragment of a token that we already output a fragment of.

    use autodie qw(:all);
    use Carp;

###########################################################################

sub new
{
    my ($class, $args) = @_;
    my $self = bless {}, $class;
    my ($char_stream, $max_frag_len)
        = @{$args}{'char_stream', 'max_frag_len'};

    confess ((caller(0))[3]).q{(): Bad :$char_stream arg;}
            . q{ it must be an open filehandle.}
        if not (ref $char_stream eq 'GLOB' and xxx $char_stream);
        # We can add support for IO::Handle and other input kinds later.

    if (!defined $max_frag_len)
    {
        $max_frag_len = $DEF_TOKEN_FRAG_LEN;
    }
    else
    {
        confess ((caller(0))[3]).q{(): Bad :$max_frag_len arg; it must be}
                . qq{ undefined or an integer >= $MIN_TOKEN_FRAG_LEN.}
            if ref $max_frag_len or not ($max_frag_len =~ m/^[0-9]+$/
                and $max_frag_len >= $MIN_TOKEN_FRAG_LEN);
    }

    binmode $char_stream, ':encoding(UTF-8)';
        # We can add support for non-Unicode(UTF-8)/ASCII input later.
        # If any input isn't valid UTF-8, we expect that the above has
        # been silently replaced with sentinel chars (codepoint 0xFFFD).

    $self->{max_frag_len}    = $max_frag_len;
    $self->{char_stream}     = $char_stream;
    $self->{char_stream_buf} = do { my $csb = q{}; \$csb; };
    $self->{is_mid_token}    = 0;

    return $self;
}

###########################################################################

sub max_frag_len
{
    return $_[0]->{max_frag_len};
}

sub update_max_frag_len
{
    my ($self) = @_;
    my ($max_frag_len) = @{$args}{'max_frag_len'};

    if (!defined $max_frag_len)
    {
        $max_frag_len = $DEF_TOKEN_FRAG_LEN;
    }
    else
    {
        confess ((caller(0))[3]).q{(): Bad :$max_frag_len arg; it must be}
                . qq{ undefined or an integer >= $MIN_TOKEN_FRAG_LEN.}
            if ref $max_frag_len or not ($max_frag_len =~ m/^[0-9]+$/
                and $max_frag_len >= $MIN_TOKEN_FRAG_LEN);
    }

    $self->{max_frag_len} = $max_frag_len;
}

###########################################################################

sub pull_token_or_fragment
{
    my ($self) = @_;
    my $payload = $self->{is_mid_token} ? $self->_continue_token()
        : $self->_start_token();
    return {
        chars        => $payload,
        is_mid_token => $self->{is_mid_token},
    };
}

###########################################################################

sub _pull_n_chars_append_into_buffer
{
    my ($self, $cnt) = @_;
    # We expect $cnt is >= 1.
    my $csb_ref = $self->{char_stream_buf};
    read $self->{char_stream}, $$csb_ref, $cnt, length $$csb_ref;
}

###########################################################################

} # class Muldis::D::RefEng::StreamLexer

###########################################################################
###########################################################################

1; # Magic true value required at end of a reusable file's code.
__END__

=pod

=encoding utf8

=head1 NAME

Muldis::D::RefEng::StreamLexer -
Process streaming Muldis D source code from characters to tokens

=head1 SYNTAX

    use Muldis::D::RefEng::StreamLexer;

    my $stream_lexer = Muldis::D::RefEng::StreamLexer->new({
        char_stream => Muldis::D::RefEng::StreamDecoder->new({
        raw_stream => \*STDIN }), max_frag_len => 150 });

    while (my $token = $stream_lexer->pull_token_or_fragment())
    {
        # And do work with $token.
    }

=head1 DESCRIPTION

Muldis::D::RefEng::StreamLexer provides a stream processor that takes a
stream of Perl characters as input and then outputs a stream of Muldis D
source code syntactical tokens or fragments thereof.

The input stream must be a regular Perl filehandle that is open for
reading, and the data must be valid ASCII or Unicode UTF-8; that is, this
module will apply a binmode of ":encoding(UTF-8)" before reading from it.
In the future, this module will be updated to support other kinds of input.

This module just splits up Muldis D source code strings into substrings,
and it does not add or remove or substitute any characters; if the returned
substrings were simply printed to an output filehandle in the same order,
the output stream/file should be identical to the input.

This module is used internally by L<Muldis::D::RefEng>, such that it serves
as a helper for L<Muldis::D::RefEng::StreamParser>, which assigns more
semantic metadata to the tokens and normalizes them.

This module generally returns whole tokens at once, as typically each one
is quite short (such as for an operator name or numeric literal), but
sometimes it returns tokens in fragments instead, such as because the whole
token would be quite long (such as for some Text or Blob literals) and we
wish to preserve working memory, either for performance or for security,
or protection against malformed input (such as an unterminated string).

Each returned token or fragment is simply a Perl character string, and no
metadata is returned with it.  Since returned tokens are rather
fine-grained (for example, each delimiter of a string literal of the input
is its own token), the user of this module should be able to tell easily
enough what kind each token is, and whether or not it is complete or a
fragment, just from said character strings themselves.

=head1 UNDERSTANDING TOKENS

Broadly speaking, every token or fragment that this module outputs will
arise from 2 disjoint and complementary contexts within the input, one of
which is anywhere outside all quoted strings, and the complement is
anywhere inside any quoted string; quoted strings don't overlap or nest, or
if they appear to, any inner ones are just regular string contents.

There are 2 main forms of quoted strings.  The first form is code comments,
whose opening delimiter is C</*> and whose closing delimiter is C<*/>, and
any occurrence of C<*/> meant to be part of the string content must be
C<\>-prefixed.  The second form is non-comments, which has a sub-form for
each of the 3 delimiter characters {C<'>,C<">,C<`>}, where for each
sub-form both the opening and closing delimiter must be the same, and any
occurrence of that delimiter character meant to be part of the string
content must be C<\>-prefixed.

This module will always return each occurrence of a string delimiter that
causes the transition to or from quoted string context as its own whole
token (never a fragment), and so every token except one of these is either
wholly inside or outside quoted string context, as is every fragment.

=head2 Outside String Context

Within normal input code, outside string context, this module will always
transition between tokens where the input character string transitions
between any 2 of the following character groups:

=over

=item B<Alphanumeric>

The ASCII letters, digits, and underscore; that is:
C<{'a'..'z','A'..'Z','0'..'9','_'}>

=item B<Standalone>

The ASCII backslash, comma, parenthesis, brackets, braces; that is:
C<{'\',',','(',')','[',']','{','}'}>.  Moreover, as a special exception,
any runs of these characters are also split into tokens as well, so any
occurrence of these will produce single-character tokens (never fragments).

=item B<Symbolic>

The printable ASCII characters not otherwise mentioned, including the
comment string delimiter characters when not in that exact sequence; that
is: C<{'!','#','$','%','&','*','+','-','.','/',':',
'<','=','>','?','@','^','|','~'}>.  Moreover, as a special exception, a run
of these characters will be split further into tokens if a C<:> is present;
specifically, any C<::=> will be its own token (never fragment), or
otherwise any C<:=> will be its own, or else any C<:> will be its own.

=item B<Whitespace>

The ASCII space, tab, linebreak, carriage return; that is: C<{'
','\t','\n','\r'}> aka C<{0x20,0x9,0xA,0xD}>.  Moreover, as a special
exception, a run of these characters will be split further into tokens if
any linebreaks are present, such that any logical line separator will be
its own token (never fragment); specifically, any C<[0xD,0xA]> pair will be
its own token, or otherwise any C<0xA> or C<0xD> will be its own.

=item B<Illegal>

All ASCII control characters besides those named in B<Whitespace> may not
appear literally in the input stream; that is: C<{0x0..0x8,0xB..0xC,
0xE..0x1F,0x7F}>.  If these are found, runs of such will be returned as
their own tokens or fragments, and the user code should complain about it.

=back

The above applies specifically when the input code declares itself
specifically to have the ASCII character repertoire.  When the input code
declares itself to be Unicode or some other repertoire instead, then these
sets of characters are extended with a useful whitelist: B<Alphanumeric>,
B<Symbolic>, B<Whitespace>; any other characters from the wider repertoire
go into B<Illegal>.  To be more specific, probably the only B<Whitespace>
addition would be the one or so line separator characters added by Unicode,
while C<Symbolic> would likely gain the most additions, at least several
dozen mathematical/etc operator symbols; B<Alphanumeric> may gain the alpha
characters of several other scripts, if it is useful to do so, but
certainly not any numerics of other scripts.

=head2 Inside String Context

Inside string context, this module will always transition between tokens
where the input has any of the following characters:

=over

=item B<Line Separator>

Any logical line separator will be its own token (never fragment);
specifically, any C<[0xD,0xA]> pair will be its own token, or otherwise any
C<0xA> or C<0xD> will be its own.

=item B<Escape Sequence>

The ASCII backslash (C<\>) where found will be returned as its own token.
Moreover, the single character immediately following it will also be
returned as its own token, except where a B<Line Saparator> immediately
follows the backslash.  Moreover, if the aforementioned single character is
C<c> then a subsequent short sequence of characters identifying the
codepoint of another character will also be returned as its own token
(possibly fragmented in degenerate cases).

=item B<Illegal>

All control characters besides a short whitelist consisting of space, tab
and line separators may not appear literally in the input stream.  If these
are found, runs of such will be returned as their own tokens or fragments,
and the user code should complain about it.  Logically including any such
characters in strings may only be done using B<Escape Sequence>.

=back

When the input code declares itself to be Unicode or some other repertoire
instead, then generally within string context any printable character may
appear literally in the input stream, while generally any non-printable or
control character may not; there would be a white-list for the
non-printable characters allowed, and any others may only be specified
using B<Escape Sequence>.

=head1 INTERFACE

The interface of Muldis::D::RefEng::StreamLexer is entirely
object-oriented; you use it by creating objects of the
Muldis::D::RefEng::StreamLexer class and then invoking methods on those
objects.  All of their attributes are private, so you must use accessor
methods.

The usual way that Muldis::D::RefEng::StreamLexer indicates a failure is to
throw an exception; most often this is due to invalid input.  If an invoked
routine simply returns, you can assume that it has succeeded, even if the
return value is undefined.

...

The C<max_frag_len> optional configuration parameter lets you tune the
maximum size of each fragment in characters to wherever it might be optimal
for performance or debugging; it defaults to 100 and must be at least 10 if
given.  By default, longer tokens will be split into fragments of the
configured size, but that as a special case, if the token contains any
line-ending characters such as C<\n> or C<\r> then it will be split
immediately following those, so that this module's user can more easily
report debugging information in terms of source code line numbers.

And so, if C<max_frag_len> is set to be larger than all source code lines,
then the only fragmentation that will occur is when any delimited strings
or comments wrap around to multiple source lines; hence the default of 100.

=head1 DEPENDENCIES

This file requires any version of Perl 5.x.y that is at least 5.8.1, and
recommends one that is at least 5.16.2.

It also requires these Perl 5 packages that are bundled with at least the
latest production version of Perl 5, or are otherwise available on CPAN for
older supported Perl 5 versions: L<autodie>, L<Carp>.

It also requires these Perl 5 packages that are in the current
distribution:
L<Muldis::D::RefEng::StreamDecoder-ver(0.0.0..*)|Muldis::D::RefEng::StreamDecoder>.

=head1 AUTHOR

Darren Duncan - darren@DarrenDuncan.net

=head1 LICENSE AND COPYRIGHT

Copyright © 2011-2015, Muldis Data Systems, Inc.

See the LICENSE AND COPYRIGHT of L<Muldis::D::RefEng> for details.

=cut


lib/Muldis/D/RefEng/StreamDecoder.pm
use 5.008001;
use utf8;
use strict;
use warnings FATAL => 'all';

###########################################################################
###########################################################################

{ package Muldis::D::RefEng::StreamDecoder; # class
    BEGIN {
        our $VERSION = '0.000000';
        $VERSION = eval $VERSION;
    }

    # ATTRIBUTE LIST:
        # raw_stream : A filehandle open for reading, we pull input from.
        # raw_stream_buf : Str (ref) : Any input we pulled from raw_stream
            # that we haven't yet normalized for output; this is
            # conceptually un-pulled but we're using it to
            # to effectively look ahead into the source stream.
        # out_stream_buf : Str (ref) : Any chars that have been normalized
            # and'r ready for output.  The pull_char method will first take
            # the first codepoint from out_stream_buf if there are any, and
            # only when out_stream_buf is empty will raw_stream/buf draw.

    use autodie qw(:all);
    use Carp;
    use Scalar::Util 'openhandle';

    use Unicode::Normalize 1.16;

###########################################################################

sub new
{
    my ($class, $args) = @_;
    my $self = bless {}, $class;
    my ($raw_stream) = @{$args}{'raw_stream'};

    confess ((caller(0))[3]).q{(): Bad :$raw_stream arg;}
            . q{ it must be an open filehandle.}
        if not (ref $raw_stream eq 'GLOB' and openhandle $raw_stream);
        # We can add support for IO::Handle and other input kinds later.

    binmode $raw_stream, ':encoding(UTF-8)';
        # We can add support for non-Unicode(UTF-8)/ASCII input later.
        # If any input isn't valid UTF-8, we expect that the above has
        # been silently replaced with sentinel chars (codepoint 0xFFFD).

    $self->{raw_stream}     = $raw_stream;
    $self->{raw_stream_buf} = do { my $rsb = q{}; \$rsb; };
    $self->{out_stream_buf} = do { my $osb = q{}; \$osb; };

    return $self;
}

###########################################################################

# MAYBE CHANGE THIS API TO ADD OFFSETS FOR PEEKING?

sub pull_char
{
    my ($self) = @_;
    $self->_extend_out_stream_buf_to_count( 1 );
    # Slice 1 char from start of out_stream_buf and return sliced.
    return substr ${$self->{out_stream_buf}}, 0, 1, q{};
}

sub pull_n_chars
{
    my ($self, $args) = @_;
    my ($count) = @{$args}{'count'};

    confess ((caller(0))[3]).q{(): Bad :$count arg; it must be}
            . qq{ an integer >= 1.}
        if !defined $count or ref $count
            or not ($count =~ m/^[0-9]+$/ and $count >= 1);

    $self->_extend_out_stream_buf_to_count( $count );
    # Slice $count chars from start of out_stream_buf and return sliced.
    return substr ${$self->{out_stream_buf}}, 0, $count, q{};
}

sub peek_char
{
    my ($self) = @_;
    $self->_extend_out_stream_buf_to_count( 1 );
    # Copy 1 char from start of out_stream_buf and return that.
    return substr ${$self->{out_stream_buf}}, 0, 1;
}

sub peek_n_chars
{
    my ($self, $args) = @_;
    my ($count) = @{$args}{'count'};

    confess ((caller(0))[3]).q{(): Bad :$count arg; it must be}
            . qq{ an integer >= 1.}
        if !defined $count or ref $count
            or not ($count =~ m/^[0-9]+$/ and $count >= 1);

    $self->_extend_out_stream_buf_to_count( $count );
    # Copy $count chars from start of out_stream_buf and return that.
    return substr ${$self->{out_stream_buf}}, 0, $count;
}

###########################################################################

# This routine makes sure that raw_stream_buf contains at least $count raw
# chars; it should only return short of that if raw_stream is exhausted.

sub _extend_raw_stream_buf_to_count
{
    my ($self, $count) = @_;
    # We expect $count is >= 1.
    my $rsb_ref = $self->{raw_stream_buf};
    if (length $$rsb_ref < $count)
    {
        read $self->{raw_stream}, $$rsb_ref,
            $count - length $$rsb_ref, length $$rsb_ref;
    }
}

###########################################################################

# This routine makes sure that out_stream_buf contains at least $count
# normalized chars; it should only return short if raw_stream is exhausted.

sub _extend_out_stream_buf_to_count
{
    my ($self, $count) = @_;
    # We expect $cnt is >= 1.
    my $osb_ref = $self->{out_stream_buf};
    if (length $$osb_ref < $count)
    {
        # See how many more normalized chars we need to procure.
        my $osb_shortfall = $count - length $$osb_ref;

        my $rsb_ref = $self->{raw_stream_buf};
        if (length $$rsb_ref < 1 + $osb_shortfall)
        {
            $self->_extend_raw_stream_buf_to_count( 1 + $osb_shortfall );
        }
        # At this point either raw_stream_buf should contain at least 1
        # more raw character than the count of normalized characters we are
        # short by, or raw_stream is exhausted.

        # Now, since normalization to NFD or NFKD will either leave alone
        # or split raw_stream_buf characters, we can guarantee that not
        # more than $osb_shortfall chars will be moved from raw_stream_buf
        # to out_stream_buf, except when the raw char immediately following
        # the last raw char that would have been moved is a combining char,
        # in which case any single or run of combining chars starting there
        # would also be moved, to ensure that all codepoint sequences in
        # out_stream_buf which are mutually combinable are in canonical
        # order; this might require further extending raw_stream_buf if so.

        # TODO, TESTS FOR CHARACTER PROPERTIES, DOING APPROP NORMALIZE ETC
        # SEE ALSO:
        # UAX 15 - 13.1 Buffering with Unicode Normalization
        # UAX 15 - 9 Detecting Normalization Forms
        # QUOTH:
#         That is, as decompositions are appended to the buffer,
#         periodically the end of the buffer will be reached. At that time,
#         the characters in the buffer up to but not including the last
#         character with the property value Quick_Check=Yes (QC=Y) must be
#         canonically ordered (and if NFC and NFKC are being generated, must
#         also be composed), and only then flushed.

        ...;

    }
}

###########################################################################

} # class Muldis::D::RefEng::StreamDecoder

###########################################################################
###########################################################################

1; # Magic true value required at end of a reusable file's code.
__END__

=pod

=encoding utf8

=head1 NAME

Muldis::D::RefEng::StreamDecoder -
Process streaming Muldis D source code from raw form to characters

=head1 SYNTAX

    use Muldis::D::RefEng::StreamDecoder;

    my $stream_decoder = Muldis::D::RefEng::StreamDecoder->new({
        raw_stream => \*STDIN });

    while (my $char = $stream_decoder->pull_char())
    {
        # And do work with $char.
    }

=head1 DESCRIPTION

Muldis::D::RefEng::StreamDecoder provides a stream processor that takes a
stream of possibly encoded characters as input and then outputs a stream of
Perl native/internal and possibly normalized characters.

This module is used internally by L<Muldis::D::RefEng>, such that it serves
as a helper for L<Muldis::D::RefEng::StreamLexer>, which groups the
characters into token strings.

The input stream must be a regular Perl filehandle that is open for
reading, and the data must be valid ASCII or Unicode UTF-8; that is, this
module will apply a binmode of ":encoding(UTF-8)" before reading from it.
In the future, this module will be updated to support other kinds of input,
meaning either non-UTF-8 filehandles or Perl string values.

This module's primary purpose is to cause the interpretation of two
arbitrary Muldis D source code strings to be identical when they differ
only by encoding or Unicode normal form.  This module outputs a character
stream where each character is a Perl native character string consisting of
a single Unicode (superset) codepoint that is noncomposed, no matter the
encoding of the input.  This module's output is always Unicode canonical
decomposed normal form (NFD).  If the returned characters were simply
printed to an output filehandle in the same order, the output stream/file
should be identical to the input, for all practical intents and purposes.

As an optional feature, this module can be configured to also fold its
input into subset character repertoire, by way of Unicode compatibility
decomposed normal form (NFKD), which may be useful for security.  Muldis D
source code is expected to declare whether it expects this compatibility
folding to occur, and if so then the user of this module should check for
this and change the configuration of this module as appropriate.

This module always performs at least Unicode canonical normalization on its
input and so if you want to specify character data that is not in this
normal form then you can not do it in Muldis D source code with literal
codepoints but rather you must use escape sequences.

This module can optionally buffer the input stream, so that after its user
gets some output, it can choose to start over and have this module
re-process the same input into possibly different output.  For example,
there might be initial runs just to scan for instructions on how to
interpret the entire stream, such as an ambiguity-resolving encoding
declaration or a request for the compatibility folding.  But even absent an
explicit buffering request, a few input characters may be buffered by this
module in order to work, and so external reading of the underlying input
filehandle is not advised as it may miss some characters.

=head1 INTERFACE

The interface of Muldis::D::RefEng::StreamDecoder is entirely
object-oriented; you use it by creating objects of the
Muldis::D::RefEng::StreamDecoder class and then invoking methods on those
objects.  All of their attributes are private, so you must use accessor
methods.

The usual way that Muldis::D::RefEng::StreamDecoder indicates a failure is to
throw an exception; most often this is due to invalid input.  If an invoked
routine simply returns, you can assume that it has succeeded, even if the
return value is undefined.

=head1 Constructor Submethods

This is currently the only routine declared by
Muldis::D::RefEng::StreamDecoder that you invoke off of the class name;
currently you invoke all other routines off of a
Muldis::D::RefEng::StreamDecoder object.

=head2 new

C<submethod new of Muldis::D::RefEng::StreamDecoder (FileHandle :$raw_stream)>

This constructor submethod creates and returns a new
C<Muldis::D::RefEng::StreamDecoder> object, which is a wrapper and buffer
for the (open and ready for reading) regular Perl filehandle given, to
provide that object's input stream, in its C<$raw_stream> argument.

=head1 Object Methods

=head2 pull_char

C<method pull_char of Str ($self:)>

This method returns the next Perl character / noncomposed Unicode codepoint
of the object's output stream, also advancing that stream so that the next
invocation of C<pull_char> or C<peek_char> will return the subsequent
character.

=head2 pull_n_chars

C<method pull_n_chars of Str ($self: PInt :$count)>

This method is a generalization of C<pull_char> that returns in a single
string the next C<$count> characters instead of just the next 1.

=head2 peek_char

C<method peek_char of Str ($self:)>

This method is like C<pull_char> but that it does I<not> advance the output
stream, so the next invocation of C<pull_char> or C<peek_char> will return
the I<same> character.

=head2 peek_n_chars

C<method peek_n_chars of Str ($self: PInt :$count)>

This method is a generalization of C<peek_char> that returns in a single
string the next C<$count> characters instead of just the next 1.

=head1 DEPENDENCIES

This file requires any version of Perl 5.x.y that is at least 5.8.1, and
recommends one that is at least 5.16.2.

It also requires these Perl 5 packages that are bundled with at least the
latest production version of Perl 5, or are otherwise available on CPAN for
older supported Perl 5 versions: L<autodie>, L<Carp>, L<Scalar::Util>.

It also requires these Perl 5 packages that are on CPAN:
L<Unicode::Normalize-ver(1.16..*)|Unicode::Normalize>.

=head1 BUGS AND LIMITATIONS

Muldis::D::RefEng::StreamDecoder depends on L<Unicode::Normalize> for its
exact behavior concerning the specifics of Unicode normal forms, and that
module in turn refers to the Perl core's Unicode database in the directory
/lib/unicore (or formerly /lib/unicode), and so the Unicode version of
normalization implemented by this module depends on your Perl's version.
This is in contrast to the latest version of the Muldis D language
specification which says that Unicode version 6.2.0 defines the behavior.
See L<Unicode::Normalize> main section "CAVEATS" for more details on this.

=head1 AUTHOR

Darren Duncan - darren@DarrenDuncan.net

=head1 LICENSE AND COPYRIGHT

Copyright © 2011-2015, Muldis Data Systems, Inc.

See the LICENSE AND COPYRIGHT of L<Muldis::D::RefEng> for details.

=cut


Makefile.PL
use 5.008001;
use utf8;
use strict;
use warnings;

use ExtUtils::MakeMaker;

# Make setting optional MakeMaker parameters more readable.
sub OPTIONAL {
   return () unless $ExtUtils::MakeMaker::VERSION ge shift;
   return @_;
}

WriteMakefile(
   NAME => 'Muldis::D::RefEng',
   VERSION => '0.0.0',
   ABSTRACT => 'Reference Implementation of Muldis D Over Perl 5',
   AUTHOR => 'Darren Duncan <darren@DarrenDuncan.net>',
   PREREQ_PM => {
      # This list is a copy of "META_ADD:[configure_|build_|]requires"
      # where the other is canonical; this list is for use by MakeMaker
      # itself, which in general wouldn't see the info in the other lists,
      # as the other lists are primarily used when generating the META.yml.
      'B' => '1.02',
      'Carp' => '1.01',
      'ExtUtils::MakeMaker' => '6.17',
      'Math::BigInt' => '1.66',
      'Scalar::Util' => '1.13',
      'Test::More' => '0.47',
   },
   OPTIONAL( '6.48',
      MIN_PERL_VERSION => '5.008001',
   ),
   OPTIONAL( '6.31',
      LICENSE => 'lgpl_3',
   ),
   OPTIONAL( '6.46',
      # Use META_ADD instead of META_MERGE so that we can remove
      # any build-time dependencies that MakeMaker will put into
      # the requires field.  (Comment From DBD::SQLite; do *we* need that?)
      META_ADD => {
         configure_requires => {
            'ExtUtils::MakeMaker' => '6.17', # bundled with Perl 5.8.1
         },
         build_requires => {
            'ExtUtils::MakeMaker' => '6.17', # bundled with Perl 5.8.1
            'Test::More' => '0.47', # bundled with Perl 5.8.1
            # These are not bundled with any version of Perl.
            # 'Carp::Always' => '0.01', # first released version
            # 'Test::Deep' => '0.106', # tests req 0.106+ w Perl 5.10.1+
         },
         requires => {
            'perl' => '5.008001', # as MIN_PERL_VERSION no affect META.yml
            'B' => '1.02', # bundled with Perl 5.8.1
            'Carp' => '1.01', # bundled with Perl 5.8.1
            'Math::BigInt' => '1.66', # bundled with Perl 5.8.1
            'Scalar::Util' => '1.13', # bundled with Perl 5.8.1
            # These are not bundled with any version of Perl.
            # 'List::MoreUtils' => '0.28', # req 0.25_02+ w Perl 5.10.1+
            # Possible:
            # Unicode::Normalize => '1.16',
            # autodie
         },
         resources => {
            homepage => 'http://www.muldis.com/',
            license => 'http://www.fsf.org/licensing/licenses/lgpl.html',
            MailingList => 'http://mm.darrenduncan.net/mailman/listinfo',
            repository => 'http://github.com/muldis/Muldis-D-Ref-Perl5',
         },
      },
   ),
);

1;
